### ISSUE TITLE
`remove-unnecessary-else` bad suggestions and false positives on thread_local macro

### ISSUE THREAD

**Issue by awused** [rust-lang/rust-analyzer#16556]
**rust-analyzer version**: rust-analyzer 0.3.1839-standalone

**rustc version**: rustc 1.76.0 (07dca489a 2024-02-04)

**relevant settings**: Unlikely any are relevant

Bad suggestions: Given this code

```rust
fn main() {
    let files = [1];

    let query = if files.is_empty() {
        return;
    } else if files.len() == 1 {
        1
    } else {
        17
    };
    print!("{query}");
}
```
remove-unnecessary-else suggests

```rust
fn main() {
    let files = [1];

    let query = if files.is_empty() {
        return;
    }
    if files.len() == 1 {
        1
    } else {
        17
    };
    print!("{query}");
}
```
which doesn't compile.

I also get false positives on `thread_local!{}` macros, but I don't have a minimal reproduction. In my projects it lints on every `thread_local!{}` call, even those as simple as `thread_local!(static SOMETHING: Cell<u32> = Cell::default());` but I don't get the lints in minimal projects.

**Comment by zjp-CN** [comment]
I'm hitting the same issue here after upgrading the nightly rustc and rust-analyzer.

`rustc 1.78.0-nightly (a84bb95a1 2024-02-13)`

`rust-analyzer 2024-02-12`

Here's the reproducible code.

```rust
use icu_segmenter::LineSegmenter;
use syntect::{
    highlighting::ThemeSet,
    parsing::SyntaxSet,
};

thread_local! {
   static SYNTHEME: (SyntaxSet, ThemeSet) = (
       SyntaxSet::load_defaults_newlines(),
       ThemeSet::load_defaults(),
   );
   static SEGMENTER: LineSegmenter = LineSegmenter::new_auto();
}
```

You can see the `remove-unnecessary-else` warnings on [rustexplorer](https://www.rustexplorer.com/b/j37elt) which utilizes the latest RA I guess.

**Comment by ShoyuVanilla** [comment]
So, this is due to RA's [`check_for_unnecessary_else`](https://github.com/rust-lang/rust-analyzer/blob/dba59970bcccfb3c6fc16ea0d0d79da875f22316/crates/hir-ty/src/diagnostics/expr.rs#L291) diagnostic; it checks whether the if-then block is never type and mark else block as unnecessary if so.
This is true for something like if "statements", but not holds for if "expressions", which has a value.
My PR fixes this and the RA build from that branch removes the false positive diagnostic for **awused**'s example code (`fn main`) and still diagnosis the true ones as in `fn foo`
![image](https://github.com/rust-lang/rust-analyzer/assets/10542892/aeff4b3d-2c59-431e-a20c-0540074aea96)
Yeah, it still marks `thread_local` as warning for **zjp-CN**'s example code, but since the `thread_local` macro is expanded as (`cargo-expand`)
```rust
const SYNTHEME: ::std::thread::LocalKey<(SyntaxSet, ThemeSet)> = {
    #[inline]
    fn __init() -> (SyntaxSet, ThemeSet) {
        (SyntaxSet::load_defaults_newlines(), ThemeSet::load_defaults())
    }
    #[inline]
    unsafe fn __getit(
        init: ::std::option::Option<&mut ::std::option::Option<(SyntaxSet, ThemeSet)>>,
    ) -> ::std::option::Option<&'static (SyntaxSet, ThemeSet)> {
        #[thread_local]
        static __KEY: ::std::thread::local_impl::Key<(SyntaxSet, ThemeSet)> = ::std::thread::local_impl::Key::<
            (SyntaxSet, ThemeSet),
        >::new();
        unsafe {
            __KEY
                .get(move || {
                    if let ::std::option::Option::Some(init) = init {
                        if let ::std::option::Option::Some(value) = init.take() {
                            return value;
                        } else if true {
                            {
                                ::core::panicking::panic_fmt(
                                    format_args!(
                                        "internal error: entered unreachable code: {0}",
                                        format_args!("missing default value"),
                                    ),
                                );
                            };
                        }
                    }
                    __init()
                })
        }
    }
    unsafe { ::std::thread::LocalKey::new(__getit) }
};
const SEGMENTER: ::std::thread::LocalKey<LineSegmenter> = {
    #[inline]
    fn __init() -> LineSegmenter {
        LineSegmenter::new_auto()
    }
    #[inline]
    unsafe fn __getit(
        init: ::std::option::Option<&mut ::std::option::Option<LineSegmenter>>,
    ) -> ::std::option::Option<&'static LineSegmenter> {
        #[thread_local]
        static __KEY: ::std::thread::local_impl::Key<LineSegmenter> = ::std::thread::local_impl::Key::<
            LineSegmenter,
        >::new();
        unsafe {
            __KEY
                .get(move || {
                    if let ::std::option::Option::Some(init) = init {
                        if let ::std::option::Option::Some(value) = init.take() {
                            return value;
                        } else if true {
                            {
                                ::core::panicking::panic_fmt(
                                    format_args!(
                                        "internal error: entered unreachable code: {0}",
                                        format_args!("missing default value"),
                                    ),
                                );
                            };
                        }
                    }
                    __init()
                })
        }
    }
    unsafe { ::std::thread::LocalKey::new(__getit) }
};
```
and this is not wrong as itself since 

```rust
                        if let ::std::option::Option::Some(value) = init.take() {
                            return value;
                        } else if true {
                        //^^^^ This `else` is unnecessary, indeed
                            {
                                ::core::panicking::panic_fmt(
                                    format_args!(
                                        "internal error: entered unreachable code: {0}",
                                        format_args!("missing default value"),
                                    ),
                                );
                            };
                        }
```
So, we should look for another approach for this `thread_local` case

**Comment by ShoyuVanilla** [comment]
> So, we should look for another approach for this thread_local case

And I've opened a PR that fixes this from `thread_local` macro implementation code
rust-lang/rust#121098

**Comment by Nilstrieb** [comment]
Happy to take the fix, but this is also an r-a bug, it shouldn't lint things from other crates like this (just like rustc does).

**Comment by awused** [comment]
Honestly, I'm not even sure the code is better after https://github.com/rust-lang/rust/pull/121098. At best it's arguable, I think the changed code can make the meaning of the code less clear.

```rust
if foo {
    return bar;
} else {
    do_something_else();
}
```

changing to

```rust
if foo {
    return bar;
}
do_something_else();
```

is fine. 

But this code snippet

```rust
if foo {
    return bar;
} else if bar {
    do_something_else();
}
```

becoming

```rust
if foo {
    return bar;
}
if bar {
    do_something_else();
}
```

makes it less obvious that the intended condition for the second branch is `!foo && bar`. This might cause confusion later during changes or refactors, not to mention it increases the line count without increasing clarity. I'm not sure if it should fire on else-if branches.

**Comment by ShoyuVanilla** [comment]
> Happy to take the fix, but this is also an r-a bug, it shouldn't lint things from other crates like this (just like rustc does).

Thanks, I'll take a look into rustc diagnostics on foreign crates' macro expansions. Maybe I could get some clues that suitable for here

**Comment by ShoyuVanilla** [comment]
> This might cause confusion later during changes or refactors, not to mention it increases the line count without increasing clarity. I'm not sure if it should fire on else-if branches.

That makes sense. I've seen many lsp/linters making warnings about similar if-else things and sometimes they are irritating and not seem to be "improved" by suggested changes.

Are there some idiomatic agreements or decisions about this within rust analyzer maintainers or rust community?

**Comment by zjp-CN** [comment]
Thanks for looking into this! @ShoyuVanilla 

The thread_local case can be minimized as follows
```rust
thread_local! {
    static A : Vec<u8> = Vec::from([1]);
}
```

I tried this snippet last night, but RA didn't report remove-unnecessary-else. At present, I tried again, this time RA does emit it  😮 (I didn't change anything about rustc and RA neither the configs) 
Maybe I've missed something...

**Comment by ShoyuVanilla** [comment]
> I tried this snippet last night, but RA didn't report remove-unnecessary-else. At present, I tried again, this time RA does emit it 😮 (I didn't change anything about rustc and RA neither the configs)
Maybe I've missed something...

Since the `thread_local!` macro is expanded as [this](https://github.com/rust-lang/rust/blob/ee9c7c940c07d8b67c9a6b2ec930db70dcd23a46/library/std/src/sys/pal/common/thread_local/fast_local.rs#L11), some thread local usages - e.g. const initializations - wouldn't be expanded as codes containing `if { return; } else { ... }`.
If your snippets in both tries are identical, this would be not the case, of course.
I suspect that the rust analyzer was busy running other things and could not show you the diagnostics in time on your previous try, but it maybe not the case, either 🤔

**Comment by ergl** [comment]
I see this issue has been closed by #16590, but I wonder if it also fixes the following, which just happened to me today:

**rust-analyzer version**: rust-analyzer 1.75.0-beta.5 (1a06ac5b5d7 2023-12-01)

**rustc version**: rustc 1.74.1 (a28077b28 2023-12-04)

Before (rewritten code just to show the faulty behaviour):

```rust
fn test() {
    let inner = || false;
    let mut to_skip = 0;
    let mut count = 0;
    while count == 0 {
        if true {
            1;
        } else if to_skip > 0 {
            to_skip += 1;
            continue;
        } else { // Warning here
            count += 1;
            if !inner() {
                return;
            }
        }
    }
}
```

After:

```rust
fn test() {
    let inner = || false;
    let mut to_skip = 0;
    let mut count = 0;
    while count == 0 {
        if true {
            1;
        } else {
            if to_skip > 0 {
                to_skip += 1;
                continue;
            }
            count += 1;
        }
    }
}
```

The call to `inner` has been removed, making the function do something else entirely. The actual fix would either be:

```rust
...
        } else {
            if to_skip > 0 {
                to_skip += 1;
                continue;
            }
            count += 1;
            if !inner() {
                return;
            }
        }
...
```

**Comment by ShoyuVanilla** [comment]
@ergl I guess so because of [this commit](https://github.com/rust-lang/rust-analyzer/pull/16590/commits/8f6e2127c3a1537298cbda11a8f261875edd6659), which is included in that PR.
Sorry but I can't test it myself 'cause I'm outside now 😅
